File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
# 다시보기
# MST 사용
import sys
# PriorityQueue보다 heapq가 빠르고 (w, s, e)의 값 들어가는 것도 확인했으므로 heapq 쓰자!
import heapq
input = sys.stdin.readline
n, m = map(int, input().split())
# 에지 정보를 저장할 힙 큐
q = []
# 유니온 파인드 리스트(대표 노드 저장 리스트)
parent = [0] * (n+1)

# 유니온 파인드용 리스트 초기화
for i in range(n+1):
    parent[i] = i

for i in range(m):
    s, e, w = map(int, input().split())
    heapq.heappush(q, (w, s, e))

# find 연산
def find(a):
    # a가 대표 노드이면 리턴 a
    if a == parent[a]:
        return a
    else:
        parent[a] = find(parent[a])    # 재귀함수 형태
        return parent[a]
    
# union 연산
def union(a, b):
    # a와 b의 대표 노드 찾기
    a = find(a)
    b = find(b)
    # 두 원소의 대표 노드끼리 연결
    if a != b:
        parent[b] = a
        
# MST 실행
useEdge = 0     # 사용한 에지 수 저장 변수
result = 0

while useEdge < n - 1:
    w, s, e = heapq.heappop(q)
    # 같은 부모가 아닌 경우에만 연결 => 사이클이 발생하면 안되므로
    if find(s) != find(e):
        union(s, e)
        # 에지의 가중치값 더하기
        result += w
        useEdge += 1

print(result)

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
	static class Edge implements Comparable<Edge>{
		int from, to, weight;

		public Edge(int from, int to, int weight) {
			super();
			this.from = from;
			this.to = to;
			this.weight = weight;
		}
		public int compareTo(Edge e) {
			return Integer.compare(this.weight, e.weight);
		}
		
	}
	static int V, E;
	static int[] parents;
	static Edge[] edgeList;

	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		V = Integer.parseInt(st.nextToken());
		E = Integer.parseInt(st.nextToken());
		edgeList = new Edge[E];
		
		for(int i = 0; i < E; i++) {
			st = new StringTokenizer(br.readLine());
			int s = Integer.parseInt(st.nextToken());
			int e = Integer.parseInt(st.nextToken());
			int c = Integer.parseInt(st.nextToken());
			edgeList[i] = new Edge(s, e, c);
		}
		
		Arrays.sort(edgeList);
		
		make();
		int result = 0;      // MST 비용
		int count = 0;       // 연결된 간선 개수
		for(Edge edge : edgeList) {  // 비용이 작은 간선 순으로 꺼내어서 처리
			if(union(edge.from, edge.to)) {
				result += edge.weight;
				if(++count == V-1) break;
			}
		}
		System.out.println(result);
	}
	
	public static void make() {
		parents = new int[V+1];
		for(int i = 0; i <= V; i++) {
			parents[i] = i;
		}
	}
	
	public static int find(int a) {
		if(a == parents[a]) return a;
		return parents[a] = find(parents[a]);
	}
	
	public static boolean union(int a, int b) {
		int aRoot = find(a);
		int bRoot = find(b);
		
		if(aRoot == bRoot) return false;
		parents[bRoot] = aRoot;
		return true;
	}
}